doc/y_cable/design_doc.md

```diff
@@ -0,0 +1,292 @@
+
+#### Proposed Solution
+
+  - Each module of the Y cable vendor can be a class (of each transceiver type) and all we need to do is instantiate the objects of these classes as class instances and these objects will provide the interface of calling the API's for the appropriate vendor Y cable.
+  - This instantiation will be done inside xcvrd, when xcvrd starts
+  - These objects basically emulate Y cable instances and whatever action/interaction needs to be done with the Y-Cable the methods of these objects would provide that
+  - each vendor in their implementation can inherit from a base class where there will be definitions for all the supported capabilities of the Y-cable.
+  - for vendors the recommended approach in case their subclass implementation does not support a method, is to set the method equal to None. This differentiates it from a method they forgot to implement. Then, the calling code should first check if the method is None before attempting to call it.
+
+For example the base class would be like this
+
+```python
+	class YCableBase(object):
+	    def __init__(self, port):
+		self.port = port
+                <function body here>
+
+	    def toggle_mux_to_torA(self, port):
+                <function body here>
+
+	    def toggle_mux_to_torB(self, port):
+                <function body here>
+
+	    def check_prbs(self, port):
+                <function body here>
+
+	    def only_credo(self, port):
+                <function body here>
+
+	    def only_amphenol(self, port):
+                <function body here>
+
+
+```
+
+For example a typical module of the vendor can be like this
+```python
+	class YCable(YCableBase):
+
+	    def __init__(self):
+                <function body here>
+
+	    def toggle_mux_to_torA(self, port):
+                <function body here>
+
+	    def toggle_mux_to_torB(self, port):
+                <function body here>
+
+	    def check_prbs(self, port):
+                <function body here>
+
+	    def only_credo(self, port):
+                <function body here>
+
+	    def only_amphenol(self, port):
+                <function body here>
+
+```
+#### Implementation details
+- Now for xcvrd to use this solution, all it has to do is instantiate the class objects defined by importing the appropriate module. This should happen when xcvrd starts or if there is a change event  (xcvrd inserted/removed)
+
+```python
+        import importlib
+        # import the vendor to module mapping
+        from sonic_y_cable import y_cable_vendor_mapping
+ 
+        # This vendor/part_name determination subroutines will either read through eeprom
+        # or else read from the TRANSCEIVER_INFO_TABLE inside state db
+
+        vendor_name = get_vendor_name(port)
+
+        part_name = get_part_name(port)
+
+        module = y_cable_vendor_mapping.mapping[vendor_name][part_name]
+
+        from sonic_y_cable import module
+
+        importlib.reload(module)
```

Comment:
This should not be `reload()`, but rather a handful of function calls in order to load a module from source. I'm planning to introduce a shared function in sonic_py_common soon. Suggest you mock it up assuming that function is in place:

```python
from sonic_py_common.general import load_module_from_source

...

module = load_module_from_source("y_cable",  "<path_to_module_obtained_from_mapping>")
```

Reply:
fixed with suggestion, do all the vendors name their module as y_cable or is it that we always load the module as y_cable ?

Reply:
No. The vendors should name their files with unique names. But in the `load_module_from_source()` line above, we will always load the file with the module name of "y_cable".

Reply:
PR here, FYI: https://github.com/Azure/sonic-buildimage/pull/7167